README
B-LispList

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it get an idea of the modules uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install


COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2005 Joshua ben Jore

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for B-LispList

0.01    Date/time
        First version, released on an unsuspecting world.



lib/B/Lisp.pm
package B::Lisp;

=head1 NAME

B::Lisp - Adds ->lisp

=head1 VERSION

Version 0.01

=cut

$VERSION = '0.01';

=head1 SYNOPSIS

 perl -MO=Lispy yourprogram.pl

 use B qw( svref_2object );
 use B::Lispy qw( lispy );
 print lispy( svref_2object( \ &foo ) );

=head1 EXPORT

Optionally, lispy() is exported.

=cut

use strict;
use Carp ('croak');
use B qw( main_root class );

=pod

=head1 FUNCTIONS

=over 4

=item lispy( $op )

Returns a lispy representation of an opcode.

=cut

sub lispy {
    my $obj = shift;

    if ( $obj->isa('B::OP') ) {
        return B::Lisp::OP->new($obj)->stringify;
    }
    elsif ( class($obj) eq 'NULL' ) {
        return;
    }
    else {
        croak("Invalid obj $obj");
    }
}

=pod

=item compile( ... )

Undocumented.

=cut

sub compile {
    my @args = @_;
    return sub {
        local $_ = lispy( main_root() );
        if ( defined wantarray ) {
            return $_;
        }
        else {
            print;
        }
    };
}

=pod

=back

=head1 AUTHOR

Joshua ben Jore jjore@cpan.org

=head1 BUGS

Please report any bugs or feature requests to
C<bug-b-lisplist@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=B-Lisp>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 ACKNOWLEDGEMENTS

=head1 COPYRIGHT & LICENSE

Copyright 2005 Joshua ben Jore, All Rights Reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

"Conversations tend to be so much more civil when there's a chance the other person might snap and kill you."



lib/B/Lisp/Op.pm
package B::Lisp::OP;

@ISA   = 'B::OP';
*lispy = \&B::Lisp::lispy;

use strict;
use B qw( class );
use Carp qw( croak );

sub quote {
    local $_ = shift;
    s/\\/\\\\/g;
    s/\"/\\\"/g;
    qq["$_"];
}

sub attr {
    my ( $self, $node, $attr ) = @_;

    my $getter = $node->can($attr)
        or return;
    return $node->$getter();
}

sub string_attr {
    my ( $self, $node, $attr ) = @_;
    my $name;
    if ( ref $attr ) {
        ( $attr, $name ) = @$attr;
    }
    else {
        $name = $attr;
    }
    local $_ = $self->attr( $node, $attr );
    defined() or return;

    return ( ":$name" => quote($_) );
}

my %flags = (
    4   => "flag-kids",
    8   => "flag-parens",
    16  => "flag-ref",
    32  => "flag-lvalue",
    64  => "flag-stacked",
    128 => "flag-special"
);
my %pmflags = (
    0x0001 => ":retaint",
    0x0002 => ":once",

    # 4 => unused
    0x0008 => ":maybe_const",
    0x0010 => ":skipwhite",
    0x0020 => ":white",
    0x0040 => ":const",
    0x0080 => ":keep",
    0x0100 => ":global",
    0x0200 => ":continue",
    0x0400 => ":eval",
    0x0800 => ":locale",
    0x1000 => ":multiline",
    0x2000 => ":singleline",
    0x4000 => ":fold",
    0x8000 => ":extended"
);
my @private = (
    [ qr/(?=)/                         => [ 128 => 'lvalue-intro' ] ],
    [ qr/^leave(?:sub(?:lv)?|write)?$/ => [ 64  => 'refcounted' ] ],
    [   qr/^assign$/ => [
            32 => 'assign-common',
            64 => 'assign-hash'
        ]
    ],
    [ qr/^sassign$/ => [ 64 => 'assign-backwards' ] ],
    [ qr/^(?=[ms])(?:match|subst(?:const)?)$/ => [ 64 => 'runtime' ] ],
    [   qr/^trans$/ => [
            1  => 'trans-from-utf',
            2  => 'trans-to-utf',
            4  => 'trans-identical',
            8  => 'trans-squash',
            16 => 'trans-delete',
            32 => 'trans-complement',
            64 => 'trans-grows'
        ]
    ],
    [ qr/^repeat$/ => [ 64 => 'repeat-dolist' ] ],
    [   qr/^(?=[arhp])(?:rv2[gs]v|[ah]elem|padsv)$/ => [
            32          => 'deref-av',
            64          => 'deref-hv',
            ( 32 | 64 ) => 'deref-sv'
        ]
    ],
    [   qr/^entersub$/ => [
            16 => 'entersub-db',
            32 => 'entersub-hastarg',
            64 => 'entersub-nomod'
        ]
    ],
    [   qr/^rv2cv$/ => [
            4   => 'entersub-inargs',
            8   => 'entersub-amper',
            128 => 'entersub-noparen'
        ]
    ],
    [ qr/^gv$/                                => [ 32 => 'early-cv' ] ],
    [ qr/^.elem$/                             => [ 16 => 'lval-defer' ] ],
    [ qr/^(?=[erg])(?:rv2.v|gvsv|enteriter)$/ => [ 16 => 'our-intro' ] ],
    [   qr/^(?=[ahrp])(?:rv2[ah]v|pad[ah]v|[ah]elem)$/ =>
            [ 8 => 'maybe-lvsub' ]
    ],

    # lower bits may have maxarg. This is handled separately
    [ qr/^targlex$/ => [ 16 => 'target-my' ] ],

    [ qr/^(?=[ei])(?:enter)?iter$/ => [ 4 => 'iter-reversed' ] ],
    [   qr/^const$/ => [
            4   => 'const-shortcircuit',
            8   => 'const-strict',
            16  => 'const-entered',
            32  => 'const-arybase',
            64  => 'const-bare',
            128 => 'const-warning'
        ]
    ],
    [ qr/^fl[io]p$/ => [ 64 => 'flip-linenum' ] ],
    [ qr/^list$/    => [ 64 => 'list-guessed' ] ],
    [ qr/^delete$/  => [ 64 => 'slice' ] ],
    [ qr/^exists$/  => [ 64 => 'exists-sub' ] ],
    [   qr/^sort$/ => [
            1  => 'sort-numeric',
            2  => 'sort-integer',
            4  => 'sort-reverse',
            8  => 'sort-inplace',
            16 => 'sort-descend'
        ]
    ],
    [ qr/^threadsv$/ => [ 64 => 'done-svref' ] ],
    [   qr/^(?=[bo])(?:backtick|open)$/ => [
            16  => 'open-in-raw',
            32  => 'open-in-crlf',
            64  => 'open-out-raw',
            128 => 'open-out-crlf'
        ]
    ],
    [   qr/^(?=[deh])(?:exit|hush|die)$/ => [
            64  => 'hush-vmsish',
            128 => 'exit-vmsish'
        ]
    ],
    [ qr/^ft...$/ => [ 2 => 'ft-access' ] ]
);

sub number_attr {
    my ( $self, $node, $attr ) = @_;
    my $name;
    if ( ref $attr ) {
        ( $attr, $name ) = @$attr;
    }
    else {
        $name = $attr;
    }
    local $_ = $self->attr( $node, $attr );
    defined() and $_ != 0
        or return;

    my @result;
    if ( 'flags' eq $attr ) {
        my $want = $_ & 3;
        if ( 0 != $want ) {
            push @result,
                (
                  1 == $want ? "want-void"
                : 2 == $want ? "want-scalar"
                : 3 == $want ? "want-list"
                : croak("Invalid WANT/GIMME $want")
                );
        }

        for my $test (
            sort { $a <=> $b }
            keys %flags
            )
        {
            push @result, $flags{$test} if $_ & $test;
        }

        if ( 2 <= @result ) {
            my $first = shift @result;
            my $last  = pop @result;
            return (
                join(
                    ' ',
                    ":flags" => "($first",
                    @result, "$last)"
                )
            );
        }
        elsif ( 1 == @result ) {
            return (":flags ($result[0])");
        }
        else {
            return ();
        }
    }
    elsif ( 'private' eq $attr ) {
        my $private = $_;
        for my $prv (@private) {
            my ( $rx, $flags ) = @$prv;
            next if not $node->name() =~ /$rx/;

            while (@$flags) {
                my ( $test, $name ) = splice @$flags, 0, 2;
                if ( $private & $test ) {
                    push @result, $name;
                }
            }
        }

        if ( 2 <= @result ) {
            my $first = shift @result;
            my $last  = pop @result;
            return join(
                ' ',
                ":private-flags"
                    => "($first",
                @result, "$last)"
            );
        }
        elsif ( 1 == @result ) {
            return join( ' ', ":private-flags" => "($result[0])" );
        }
        else {
            return ();
        }
    }
    elsif ( 'pmflags' eq $attr ) {

        for my $test (
            sort { $a <=> $b }
            keys %pmflags
            )
        {
            push @result, $pmflags{$test}, if $_ & $test;
        }

        if ( 2 <= @result ) {
            my $first = shift @result;
            my $last  = pop @result;
            return (
                join ' ',
                ":regexp-flags" => "($first",
                @result, "$last"
            );
        }
        elsif ( 1 == @result ) {
            return ( ":regexp-flags" => "($result[0])" );
        }
        else {
            return ();
        }
    }
    elsif ( $attr =~ /^(?:arybase|cop_seq|line|seq|targ)$/ ) {

        # Pure numeric
        return (":$name $_");
    }
    else {
        croak("Invalid numeric attribute $name=$_");
    }

    croak("Invalid numeric attribute $attr");
}

sub sv_attr {
    my ( $self, $node, $attr ) = @_;
    local $_ = $self->attr( $node, $attr );
    defined() or return;

    if ( $attr eq 'warnings' ) {

        # This, from Deparse.pm.
        if ( class($_) eq 'SPECIAL' ) {
            if ( 4 == $$_ ) {
                return ( ":warnings" => "WARN_ALL" );
            }
            elsif ( 5 == $$_ ) {
                return ( ":warnings" => "WARN_NONE" );
            }
            elsif ( 6 == $$_ ) {
                return ( ":warnings" => "WARN_STD" );
            }
            else {
                croak("Invalid WARNINGs bits: $$_");
            }
        }
        elsif ( $_->isa('B::NV') ) {
            return ( ":warnings" => $_->NV );
        }
        elsif ( $_->isa('B::IV') ) {
            return ( ":warnings" => $_->IV );
        }
        else {
            croak("Invalid warnings object isn\'t an IV or NV");
        }
    }
    elsif ( 'sv' eq $attr ) { }
    elsif ( 'gv' eq $attr ) { }
    elsif ( 'io' eq $attr ) { }
    elsif ('pmreplroot')  { }
    elsif ('pmreplstart') { }
    else {
        return ( ":$attr" => $_ );
    }
}

sub stringify {
    my $self = shift;
    my $node = $$self;

    # sv warnings io gv pmreplroot pmreplstart
    my $name = $node->name();

    my @attr;
    push @attr, (
        map $self->string_attr( $node, $_ ),
        qw( pmregexp label file pmregexp precomp pmoffset )
    );
    push @attr, $self->string_attr( $node, [ 'stashpv', 'package' ] );
    push @attr, $self->number_attr( $node, 'private' );

    no warnings 'uninitialized';
    my $first = ""
        . (
        $node->can('first')
        ? lispy( $node->first() )
        : ''
        );
    my $sibling = ""
        . (
        $node->sibling->can('name')
        ? lispy( $node->sibling() )
        : ''
        );

    my $this = join( ' ',
        grep length(),
        "(" . join( ' ', grep length(), $name, @attr, $first ) . ")",
        $sibling );

    #    print "====$this====\n" if $this =~ /private/;
    return $this;
}

sub new {
    my ( $package, $node ) = @_;
    my $type  = class($node);
    my $class = $package;       # . "::" . $type;

    bless \$node, $class;
}

qq["Well --- me!" "A ----ing wizard.  I hate ----ing wizards!!"  "Well you shouldn't ---- them, then."]


lib/synthetic.lisp
'(perl_leave :desc "block exit" 
 :targ 1 :type 178 :seq 8773 :flags 13 :private 64 :children 
(perl_enter :desc "block entry" 
 :type 177 :seq 8728) 
(perl_nextstate :desc "next statement" 
 :type 174 :seq 8729 :flags 1 :private 2) 
(perl_require :desc "require" 
 :type 309 :seq 8731 :flags 6 :private 1 :children 
(perl_const :desc "constant item" 
 :type 5 :seq 8730 :flags 2 :private 64)) 
(perl_nextstate :desc "next statement" 
 :type 174 :seq 8732 :flags 1 :private 2) 
(perl_sassign :desc "scalar assignment" 
 :type 36 :seq 8735 :flags 69 :private 2 :children 
(perl_const :desc "constant item" 
 :type 5 :seq 8733 :flags 2) 
(perl_null :desc "null operation" 
 :targ 15 :flags 182 :private 3 :children 
(perl_gvsv :desc "scalar variable" 
 :type 6 :seq 8734 :flags 2))) 
(perl_nextstate :desc "next statement" 
 :type 174 :seq 8736 :flags 1 :private 2) 
(perl_sassign :desc "scalar assignment" 
 :type 36 :seq 8739 :flags 69 :private 2 :children 
(perl_const :desc "constant item" 
 :type 5 :seq 8737 :flags 2) 
(perl_null :desc "null operation" 
 :targ 15 :flags 182 :private 19 :children 
(perl_gvsv :desc "scalar variable" 
 :type 6 :seq 8738 :flags 2 :private 16))) 
(perl_nextstate :desc "next statement" 
 :type 174 :seq 8740 :flags 1 :private 2) 
(perl_sassign :desc "scalar assignment" 
 :type 36 :seq 8743 :flags 69 :private 2 :children 
(perl_const :desc "constant item" 
 :type 5 :seq 8741 :flags 2) 
(perl_null :desc "null operation" 
 :targ 15 :flags 182 :private 19 :children 
(perl_gvsv :desc "scalar variable" 
 :type 6 :seq 8742 :flags 2 :private 16))) 
(perl_nextstate :desc "next statement" 
 :type 174 :seq 8744 :flags 1 :private 2) 
(perl_sassign :desc "scalar assignment" 
 :type 36 :seq 8747 :flags 69 :private 2 :children 
(perl_qr :desc "pattern quote (qr//)" 
 :type 32 :seq 8745 :flags 2 :private 64 :children) 
(perl_null :desc "null operation" 
 :targ 15 :flags 182 :private 19 :children 
(perl_gvsv :desc "scalar variable" 
 :type 6 :seq 8746 :flags 2 :private 16))) 
(perl_nextstate :desc "next statement" 
 :type 174 :seq 8748 :flags 1 :private 2) 
(perl_list :desc "list" 
 :type 141 :seq 8753 :flags 45 :private 128 :children 
(perl_pushmark :desc "pushmark" 
 :type 3 :seq 8749 :flags 33 :private 128) 
(perl_null :desc "null operation" 
 :targ 15 :flags 5 :private 19 :children 
(perl_gvsv :desc "scalar variable" 
 :type 6 :seq 8750 :flags 2 :private 16)) 
(perl_null :desc "null operation" 
 :targ 15 :flags 5 :private 19 :children 
(perl_gvsv :desc "scalar variable" 
 :type 6 :seq 8751 :flags 2 :private 16)) 
(perl_null :desc "null operation" 
 :targ 15 :flags 5 :private 19 :children 
(perl_gvsv :desc "scalar variable" 
 :type 6 :seq 8752 :flags 2 :private 16))) 
(perl_nextstate :desc "next statement" 
 :type 174 :seq 8754 :flags 1 :private 2) 
(perl_sassign :desc "scalar assignment" 
 :type 36 :seq 8760 :flags 69 :private 2 :children 
(perl_entersub :desc "subroutine entry" 
 :targ 7 :type 166 :seq 8758 :flags 70 :private 34 :children 
(perl_pushmark :desc "pushmark" 
 :type 3 :seq 8755 :flags 2) 
(perl_const :desc "constant item" 
 :type 5 :seq 8756 :flags 34 :private 64) 
(perl_method_named :desc "method with known name" 
 :type 350 :seq 8757 :flags 2)) 
(perl_null :desc "null operation" 
 :targ 15 :flags 182 :private 3 :children 
(perl_gvsv :desc "scalar variable" 
 :type 6 :seq 8759 :flags 2))) 
(perl_nextstate :desc "next statement" 
 :type 174 :seq 8761 :flags 1 :private 2) 
(perl_sassign :desc "scalar assignment" 
 :type 36 :seq 8766 :flags 69 :private 2 :children 
(perl_entersub :desc "subroutine entry" 
 :targ 9 :type 166 :seq 8764 :flags 70 :private 35 :children 
(perl_null :desc "null operation" 
 :targ 141 :flags 6 :children 
(perl_pushmark :desc "pushmark" 
 :type 3 :seq 8762 :flags 2) 
(perl_null :desc "null operation" 
 :targ 17 :flags 6 :private 3 :children 
(perl_gv :desc "glob value" 
 :type 7 :seq 8763 :flags 2 :private 32)))) 
(perl_padsv :desc "private variable" 
 :targ 8 :type 9 :seq 8765 :flags 178 :private 128)) 
(perl_nextstate :desc "next statement" 
 :type 174 :seq 8767 :flags 1 :private 2) 
(perl_entersub :desc "subroutine entry" 
 :targ 10 :type 166 :seq 8772 :flags 69 :private 35 :children 
(perl_null :desc "null operation" 
 :targ 141 :flags 4 :children 
(perl_pushmark :desc "pushmark" 
 :type 3 :seq 8768 :flags 2) 
(perl_null :desc "null operation" 
 :targ 15 :flags 38 :private 3 :children 
(perl_gvsv :desc "scalar variable" 
 :type 6 :seq 8769 :flags 2)) 
(perl_padsv :desc "private variable" 
 :targ 8 :type 9 :seq 8770 :flags 35) 
(perl_null :desc "null operation" 
 :targ 17 :flags 6 :private 3 :children 
(perl_gv :desc "glob value" 
 :type 7 :seq 8771 :flags 2 :private 32)))))